s3torchconnector/src/s3torchconnector/_s3dataset_common.py

```diff
@@ -42,22 +38,18 @@ def parse_s3_uri(uri: str) -> Tuple[str, str]:
 
 def get_objects_from_uris(
     object_uris: Union[str, Iterable[str]], client: S3Client
-) -> Iterable[S3Reader]:
+) -> Iterable[S3BucketKey]:
     if isinstance(object_uris, str):
         object_uris = [object_uris]
     # TODO: We should be consistent with URIs parsing. Revise if we want to do this upfront or lazily.
-    bucket_key_pairs = [parse_s3_uri(uri) for uri in object_uris]
+    bucket_key_pairs = [S3BucketKey(*parse_s3_uri(uri)) for uri in object_uris]
 
-    return bucket_key_pairs_to_objects(bucket_key_pairs, client)
+    return bucket_key_pairs
 
 
-def bucket_key_pairs_to_objects(
-    bucket_key_pairs: List[Tuple[str, str]], client: S3Client
-):
-    for bucket, key in bucket_key_pairs:
-        yield client.get_object(bucket, key)
-
-
-def list_objects_from_prefix(s3_uri: str, client: S3Client) -> S3BucketIterable:
+def list_objects_from_prefix(s3_uri: str, client: S3Client) -> Iterable[S3BucketKey]:
     bucket, prefix = parse_s3_uri(s3_uri)
-    return S3BucketIterable(client, bucket, prefix)
+    s3objects = S3BucketIterable(client, bucket, prefix)
+    bucket_key_pairs = (S3BucketKey(obj.bucket, obj.key) for obj in s3objects)
+
+    return bucket_key_pairs
```

Comment:
This is unintentionally changing the behaviour of this method. This was supposed to return an iterable but currently returning a list. The issue is not with this PR but actual implementation in the first place. Lets have another look at this part. This would change the logic of S3IterableDataset.